cmd/thanos/receive.go

```diff
@@ -193,29 +193,28 @@ func runReceive(
 	{
 		// TSDB.
 		cancel := make(chan struct{})
-		// Before actually starting, we need to make sure
-		// the WAL is flushed.
-		startTimeMargin := int64(2 * time.Duration(tsdbCfg.MinBlockDuration).Seconds() * 1000)
-		if err := db.Open(); err != nil {
-			return errors.Wrap(err, "opening storage")
-		}
-		if err := db.Flush(); err != nil {
-			return errors.Wrap(err, "flushing storage")
-		}
 		g.Add(func() error {
 			defer close(dbReady)
 			defer close(uploadC)
 
+			// Before actually starting, we need to make sure the
+			// WAL is flushed. The WAL is flushed after the
+			// hashring ring is loaded.
+			startTimeMargin := int64(2 * time.Duration(tsdbCfg.MinBlockDuration).Seconds() * 1000)
+			if err := db.Open(); err != nil {
```

Comment:
With this change, the `db` instance is never used outside of this run group. Let's move the definition of the db instance into this block to make it more local.